capa/features/extractors/binexport2/arch/arm/insn.py

```diff
@@ -0,0 +1,232 @@
+from capa.features.extractors.binexport2.helpers import (
+    mask_immediate,
+    is_address_mapped,
+    get_operand_expressions,
+    get_instruction_mnemonic,
+    get_instruction_operands,
+    get_operand_register_expression,
+    get_operand_immediate_expression,
+)
+from capa.features.extractors.binexport2.binexport2_pb2 import BinExport2
+from capa.features.extractors.binexport2.arch.arm.helpers import is_stack_register_expression
+
+logger = logging.getLogger(__name__)
+
+
+def extract_insn_number_features(
+    fh: FunctionHandle, _bbh: BBHandle, ih: InsnHandle
+) -> Iterator[Tuple[Feature, Address]]:
+    fhi: FunctionContext = fh.inner
+    ii: InstructionContext = ih.inner
+
+    be2: BinExport2 = fhi.ctx.be2
+
+    instruction_index: int = ii.instruction_index
+    instruction: BinExport2.Instruction = be2.instruction[instruction_index]
+
+    if len(instruction.operand_index) == 0:
+        # skip things like:
+        #   .text:0040116e leave
+        return
+
+    mnemonic: str = get_instruction_mnemonic(be2, instruction)
+
+    if mnemonic in ("add", "sub"):
+        assert len(instruction.operand_index) == 3
+
+        expression1: Optional[BinExport2.Expression] = get_operand_register_expression(
+            be2, be2.operand[instruction.operand_index[1]]
+        )
+        if expression1 and is_stack_register_expression(be2, expression1):
+            # skip things like:
+            # add x0,sp,#0x8
+            return
+
+    for i, operand_index in enumerate(instruction.operand_index):
+        operand: BinExport2.Operand = be2.operand[operand_index]
+
+        immediate_expression: Optional[BinExport2.Expression] = get_operand_immediate_expression(be2, operand)
+        if not immediate_expression:
+            continue
+
+        value: int = mask_immediate(fhi.arch, immediate_expression.immediate)
+        if is_address_mapped(be2, value):
+            continue
+
+        yield Number(value), ih.address
+        yield OperandNumber(i, value), ih.address
+
+        if mnemonic == "add":
+            if i == 2:
+                if 0 < value < MAX_STRUCTURE_SIZE:
+                    yield Offset(value), ih.address
+                    yield OperandOffset(i, value), ih.address
+
+
+def extract_insn_offset_features(
+    fh: FunctionHandle, bbh: BBHandle, ih: InsnHandle
+) -> Iterator[Tuple[Feature, Address]]:
+    fhi: FunctionContext = fh.inner
+    ii: InstructionContext = ih.inner
+
+    be2: BinExport2 = fhi.ctx.be2
+    instruction: BinExport2.Instruction = be2.instruction[ii.instruction_index]
+
+    if len(instruction.operand_index) == 0:
+        # skip things like:
+        #   .text:0040116e leave
+        return
+
+    mnemonic: str = get_instruction_mnemonic(be2, instruction)
+    value: Optional[int] = None
+    value_index: Optional[int] = None
+
+    operands: List[BinExport2.Operand]
+    immediate_expression: Optional[BinExport2.Expression]
+
+    if mnemonic.startswith(("ldr", "str")):
```

Comment:
following up on the function sketch above, we may want a more general `match_expression_tree` routine in `binexport2.helpers` that takes a expression tree pattern and can validate some nodes and extract information from others. So you should be able to do things like:

```py
# [x1, 8]
values: Dict[str, Any] = match_expression_tree(
  expression1,
  (DEREF, (
    (OPERATOR_PLUS, (
      REGISTER_NOT_STACK,
      {"the_int": IMMEDIATE_INT})))
)

assert values["the_int"] == 8
```

then we can throw a bunch of patterns against each expression tree and extract the best match. adding new patterns should be trivial, as should finding the cases that we don't handle yet (via `logger.warning` calls).

Reply:
I absolutely agree that this is brittle, as you mentioned, we have to work within the BinExport2 data model while accounting for differences between the underlying tools (Ghidra, IDA, etc.). I like the idea of writing abstractions for the operand parsing, assuming the time commit isn't too large (we only need operand-level parsing for capa's `number` and `offset` features). The more we can abstract from the extractor code the better so I'd lean towards an abstraction at the operand level (your first example) versus the expression tree level (your second example), simply because the expression tree level still requires an understanding of how the expression tree is constructed, which in my experience has been different between the underlying tools, especially with the recent changes to the Ghidra BinExport extension. Let's discuss more if I'm missing the benefits of expression tree level abstraction. @williballenthin if you have cycles and can start an implementation I'm happy to review and extend, otherwise I won't be able to look at this until later next week.